src/Map/MapRasterizer_P.cpp

```diff
@@ -488,17 +480,144 @@ void OsmAnd::MapRasterizer_P::rasterizePolyline(
+
+                    // fix corner shifts
+                    if (originalPoints.size() >= kLastPointCnt)
+                    {
+                        // calculation of the angle between two outermost segments of the path.
+                        auto vecADir = (originalPoints[0] - originalPoints[1]).normalized();
+                        auto vecBDir = (originalPoints[2] - originalPoints[1]).normalized();
+                        auto angle = atan2(vecBDir.y, vecBDir.x) - atan2(vecADir.y, vecADir.x);
+                        if (angle > M_PI) {
+                            angle -= 2 * M_PI;
+                        } else if (angle <= -M_PI) {
+                            angle += 2 * M_PI;
+                        }
+                        // calculation of the angle between two outermost segments of the path.
+
+                        auto ctang = 1.0f/tan(angle/2.0f);
+                        if (!(ctang > 0.0f && dir > 0) && !(ctang < 0.0f && dir < 0))
+                        {
+                            Utilities::resizeVector(shiftedPoints[2], shiftedPoints[1], offset * ctang * dir);
+                        }
+                        else
+                        {
+                            // calculate additional point for corner
+                            auto additionalNormal = Utilities::computeNormalToLine(originalPoints[1], originalPoints[0], dir);
+                            auto additionalPt = originalPoints[1] + additionalNormal * offset;
+                            shiftedPoints.insert(shiftedPoints.begin() + 1, additionalPt);
+
+                            // add additional offset for corner shifted points.
+                            Utilities::resizeVector(shiftedPoints[0], shiftedPoints[1], offset/2.0f);
+                            Utilities::resizeVector(shiftedPoints[3], shiftedPoints[2], offset/2.0f);
+                            // add additional ofset for corner shifted points.
+                        }
+                    }
+                }
+                else
+                {
+                    path.lineTo(tempVertex.x, tempVertex.y);
+                }
                 intersect = true;
             }
         }
         prevCross = cross;
         pVertex = vertex;
     }
 
+    if (shiftedPoints.size() > 1)
+    {
+        Utilities::resizeVector(*(shiftedPoints.begin() + 1), *shiftedPoints.begin(), -1.4*offset);
+        Utilities::resizeVector(*(shiftedPoints.rbegin() + 1), *shiftedPoints.rbegin(), -1.4*offset);
+        // pop very first pushed point that already drawed
+        auto pt = shiftedPoints.back();
+        path.moveTo(pt.x, pt.y);
+        shiftedPoints.pop_back();
+        while (shiftedPoints.size() > 0)
+        {
+            auto pt = shiftedPoints.back();
+            path.lineTo(pt.x, pt.y);
+            shiftedPoints.pop_back();
+        }
+    }
+
+    return intersect;
+}
+
+void OsmAnd::MapRasterizer_P::drawLayer(
```

Comment:
method name is confusing... drawLineLayer?

Reply:
Updated